libbeat/keystore/file_keystore.go

```diff
@@ -0,0 +1,380 @@
+	if err != nil {
+		return false
+	}
+	f.Close()
+	return true
+}
+
+// doSave lock/unlocking of the ressource need to be done by the caller.
+func (k *FileKeystore) doSave(override bool) error {
+	if k.dirty == false {
+		return nil
+	}
+
+	temporaryPath := fmt.Sprintf("%s.tmp", k.Path)
+
+	w := new(bytes.Buffer)
+	jsonEncoder := json.NewEncoder(w)
+	if err := jsonEncoder.Encode(k.secrets); err != nil {
+		return fmt.Errorf("cannot serialize the keystore before saving it to disk: %v", err)
+	}
+
+	encrypted, err := k.encrypt(w)
+	if err != nil {
+		return fmt.Errorf("cannot encrypt the keystore: %v", err)
+	}
+
+	flags := os.O_RDWR | os.O_CREATE
+	if override {
+		flags |= os.O_TRUNC
+	} else {
+		flags |= os.O_EXCL
+	}
+
+	f, err := os.OpenFile(temporaryPath, flags, filePermission)
+	if err != nil {
+		return fmt.Errorf("cannot open file to save the keystore to '%s', error: %s", k.Path, err)
+	}
+
+	f.Write(version)
+	base64Encoder := base64.NewEncoder(base64.StdEncoding, f)
+	io.Copy(base64Encoder, encrypted)
+	base64Encoder.Close()
+	f.Sync()
+	f.Close()
+
+	err = file.SafeFileRotate(k.Path, temporaryPath)
+	if err != nil {
+		os.Remove(temporaryPath)
+		return fmt.Errorf("cannot replace the existing keystore, with the new keystore file at '%s', error: %s", k.Path, err)
+	}
+	os.Remove(temporaryPath)
+
+	k.dirty = false
+	return nil
+}
+
+func (k *FileKeystore) load() error {
+	k.Lock()
+	defer k.Unlock()
+
+	f, err := os.OpenFile(k.Path, os.O_RDONLY, filePermission)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil
+		}
+		return err
+	}
+	defer f.Close()
+
+	if common.IsStrictPerms() {
+		if err := k.checkPermissions(k.Path); err != nil {
+			return err
+		}
+	}
+
+	raw, err := ioutil.ReadAll(f)
+	if err != nil {
+		return err
+	}
+
+	v := raw[0:len(version)]
+	if !bytes.Equal(v, version) {
+		return fmt.Errorf("keystore format doesn't match expected version: '%s' got '%s'", version, v)
+	}
+
+	base64Content := raw[len(version):]
+	if len(base64Content) == 0 {
+		return fmt.Errorf("corrupt or empty keystore")
+	}
+
+	base64Decoder := base64.NewDecoder(base64.StdEncoding, bytes.NewReader(base64Content))
+	plaintext, err := k.decrypt(base64Decoder)
+	if err != nil {
+		return fmt.Errorf("could not decrypt the keystore: %v", err)
+	}
+	jsonDecoder := json.NewDecoder(plaintext)
+	return jsonDecoder.Decode(&k.secrets)
+}
+
+// Encrypt the data payload using a derived keys and the AES-256-GCM algorithm.
+func (k *FileKeystore) encrypt(reader io.Reader) (io.Reader, error) {
+	// randomly generate the salt and the initialization vector, this information will be saved
+	// on disk in the file as part of the header
+	iv, err := common.RandomBytes(iVLength)
+
+	if err != nil {
+		return nil, err
+	}
+
+	salt, err := common.RandomBytes(saltLength)
+	if err != nil {
+		return nil, err
+	}
+
+	// Stretch the user provided key
+	password, _ := k.password.Get()
+	passwordBytes := pbkdf2.Key(password, salt, iterationsCount, keyLength, sha512.New)
```

Comment:
Consider replacing pbkdf2 with a never key derivation function like scrypt or even Argon2. Argon2 was very recently merged into golang.org/x/crypto/ .
Some resources about the topic:
https://www.linkedin.com/pulse/top-password-hashing-schemes-employ-today-chintan-jain-mba-cissp
https://download.libsodium.org/doc/password_hashing/
https://core.trac.wordpress.org/ticket/39499
https://gitlab.com/cryptsetup/cryptsetup/issues/119

In addition I would recommend to store the key derivation function and the iteration along side with the VERSION|SALT|IV|PAYLOAD.
If you decide to change the iteration count or the key derivation function you don't need to increase the file format version.
This is kind of similar to the https://secure.php.net/manual/en/function.password-hash.php. In your case the the key would be left out.

With this implementation it's possible to increase the key derivation parameters to prevent faster CPU/GPU's. To achieve this the maintainer needs to alter the key derivation parameters from time to time and if a change is detected from the key derivation parameters stored in the storage file the encrypted payload needs to be decrypted and encrypted with the new parameters.
Similar to https://secure.php.net/manual/en/function.password-needs-rehash.php

To reduce code duplication I would extract the passwordBytes generation `passwordBytes := pbkdf2.Key(password, salt, iterationsCount, keyLength, sha512.New)` to it own private function.

Btw: The rest of the encryption part looks good to me.

Reply:
@dol I've created the following issues:

https://github.com/elastic/beats/issues/6015
https://github.com/elastic/beats/issues/6016

I will do the change right away to add the function in the format and check for argon2.